### PR TITLE
Make IpvXAddr::new const fns and the well known addresses associated constants

### LINKED ISSUE #44582: Tracking issue for IP constructors
Tracking issue for IP address convenience constructors, added in #44395, modified in https://github.com/rust-lang/rust/pull/52872.

```rust
impl Ipv4Addr {
    pub const LOCALHOST: Self = Ipv4Addr::new(127, 0, 0, 1);
    pub const UNSPECIFIED: Self = Ipv4Addr::new(0, 0, 0, 0);
    pub const BROADCAST: Self = Ipv4Addr::new(255, 255, 255, 255);
}

impl Ipv6Addr {
    pub const LOCALHOST: Self = Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1);
    pub const UNSPECIFIED: Self = Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0);
}

### PR DESCRIPTION
**Pull request by faern**
Implements/fixes https://github.com/rust-lang/rust/issues/44582

I just got a PR towards libc (https://github.com/rust-lang/libc/pull/1044) merged. With the new feature added in that PR it is now possible to create `in6_addr` instances as consts. This enables us to finally make the constructors of the IP structs const fns and to make the localhost/unspecified addresses associated constants, as agreed in the above mentioned tracking issue.

I also added a BROADCAST constant. Personally this is the well known address I tend to need the most often.

### CODE REVIEW

**Comment by faern** [comment]
I realize now we might want to use the same feature gate name for the associated constants and the constructors. I did not think about that before. But I'll wait for some feedback before doing anything about it.

**Comment by faern** [comment]
It should be noted that we can get away without unlocking `#![feature(const_int_ops)]` if desired. We could have a local `const fn hton(u32) -> u32` that takes care of converting to big endian.

We can also get the associated constants without making the constructors `const fn`s. We can just have private const constructors that we use in the normal constructors and for creating the associated constants, but the public constructors can be non-const if that is desired for some reason. I went with making the constructors const since I think that would be very useful in general. Having the IP constructors const is what drove me to implement this :)

**Comment by TimNN** [comment]
Thanks for your PR! This looks generally good, I'll take a detailed look tomorrow. Having an unstable const-constructor and using `const_int_ops` is fine, I think.

**Comment by TimNN** [comment]
@bors retry

* travis-ci/travis-ci#4924

**Comment by faern** [comment]
Hmm. Looks like redox for some reason uses libc without the align feature. Not sure why. Is there any other place where std depends on libc than in `libstd/Cargo.toml`? I now tried activating the "align" feature by default in the `libc_shim` crate, but I don't know if it will help.

File: src/libstd/Cargo.toml [hunk 1]
```diff
@@ -19,7 +19,7 @@ alloc_system = { path = "../liballoc_system" }
 panic_unwind = { path = "../libpanic_unwind", optional = true }
 panic_abort = { path = "../libpanic_abort" }
 core = { path = "../libcore" }
-libc = { path = "../rustc/libc_shim" }
+libc = { path = "../rustc/libc_shim", features = ["align"] }
```
Comment (TimNN) [inline src/libstd/Cargo.toml]: Since you made `align` a default feature now, I think you can remove it here.

**Comment by faern** [comment]
@TimNN Done. Not sure if it fixes anything though, but we'll see I guess.

**Comment by TimNN** [comment]
Alright, let's try this again.

@bors r+

**Comment by faern** [comment]
Not sure what causes this :thinking:. First I thought I might have done a mistake in libc so that the conditional compilation for `__align` was wrong on redox. But the condition there is very simple, and from what I can see identical to the other platforms.

Might there be something special going on for certain platforms in this repository? Maybe pulling in libc from crates.io in certain setups? I fail to see any way libstd might depend on libc other than the place I have already tried. Maybe @alexcrichton have some insights?

**Comment by faern** [comment]
I got a merge conflict over `src/Cargo.lock`. I solved that by rebasing everything and in the process also squashing together some commits that basically did the same thing over and over. This makes following the changes harder here on github and I'm sorry if this was not an appropriate solution.

Apart from bumping libc to 0.2.43 in `Cargo.lock` I also found I could simplify the `Ipv6Addr::from(octets)` implementation to not use `unsafe` any more as well. So now there are zero unsafes in the `net/ip.rs` module :tada:

**Comment by faern** [comment]
I tried getting the `dist-various-1` build to build on my own travis account. So I could iterate and debug. But for normal OSS builds the job is terminated after ~50 minutes, and this job takes longer. So I can't debug really.

I guess I could do some docker setup and reproduce the build locally. But have not had the time to look into that yet. Anyway, there are changes here that *might* work. So if you want to you can try to merge it again and we'll see what rust-highfive has to say :)

**Comment by TimNN** [comment]
Thanks, @faern! The GitHub Ui says there is another conflict in `Cargo.lock`, could you take another look?

**Comment by faern** [comment]
@TimNN Done. Simply rebased again. I know this is not preferred from a reviewer standpoint, and unless you say it's fine I'll use another approach if this becomes an issue again.

**Comment by TimNN** [comment]
Thanks, let's try this again. (Also, rebasing is totally fine towards the end of the PR / if there are merge conflicts).

@bors r+

**Comment by faern** [comment]
@TimNN There, I found it with the help of some Redox people on IRC. Apparently Redox does not use `in6_addr` from libc, but rather define their own version deep down in libstd. This should hopefully build now!

**Comment by TimNN** [comment]
Awesome work, @faern!

@bors r+

**Comment by faern** [comment]
Trying to get anything merged in `Cargo.lock` is a pain :sweat_smile: And I strictly don't need that any more, since the fix was to change `in6_addr` in libstd itself. So if it fails one more time because of a merge conflict I'll skip trying to bump libc in Cargo.lock.

**Pull request merged**